subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/DefaultGroovyJavaJointCompileSpec.java

```diff
@@ -46,4 +47,14 @@ public void setGroovyClasspath(List<File> groovyClasspath) {
     public boolean incrementalCompilationEnabled() {
         return getCompileOptions().getPreviousCompilationDataFile() != null;
     }
+
+    @Override
+    public File getJavaDestinationDir() {
```

Comment:
Yes, this are java classes destination dir for joint compilation. Previously we just added `spec.getDestinationDir()` to Groovy compiler classpath, since we assume that Java output is in the same folder as Groovy. 

But since we modify this path before compilation that is not true anymore.

So previously we had `build/classes` as `destinationDir`, and:
1. we set `destinationDir` to `build/classes` at start
2. we added `destinationDir (`build/classes`) to classpath
3. we compile

Now this is a bit different, and we do:
1. we set `destinationDir` to `build/classes` at start|
2. we set `javaDestinationDir` to `build/classes` at start
3. for incremental compilation we change `destinationDir` to let's say `temp/classes`
4. we add `javaDestinationDir (`build/classes`) to classpath (so same folder as before)
5. we compile
6. we revert `destinationDir` to `build/classes`

Reply:
I don't understand what is going on. If we do joint compilation, then there is only one compile task and that should put all the output in one place, right? What happens if Java compilation fails as part of joint incremental Groovy compilation? Will we then again be in a bad state possibly?

If we have separate compilation (aka two compile tasks, one for Java and one for Groovy), what happens then? Is `Spec.getJavaDestinationDir()` important there as well?

Reply:
I will have to double check what is going on with joint compilation and maybe add some test

Reply:
It seems that we don't need this, so I removed it and add some tests that mix source sets still work.